### PR TITLE
Fix detection and handling of POD types (on master)

### PR DESCRIPTION
**Pull request by mppf**
The compiler would like to optimize to use bit-copies for types when it is
legal and to use no-init/assign when it is legal.  In order to do so correctly,
this patch introduces an idea of POD types. A record type or a tuple type is a
POD type if:
 * it does not set pragma "ignore noinit"
 * it has only a compiler generated autoCopy
 * it has only a compiler generated autoDestroy
 * it has only a compiler generated initialization/construction routine
 * it has only a compiler generated = routine (when LHS.type=RHS.type)
 * it contains only POD type fields.

Numeric, boolean, and pointer types (such as _ddata or c_ptr)
are always POD. User class types are always POD.
Sync/single types are never POD. Atomic types are never POD.

For POD types,  we can
 * shallow/bit copy them across locales without calling overrides
 * avoid ever calling autoCopy/autoDestroy
 * replace record assignment with bit copy/ C = operator.

This patch reworks bulkCopyRecords to use the idea of a POD type since
previously, it would incorrectly think an assignment overload like

     lhs.x = rhs.y;
     rhs.y = rhs.x;

would be trivial.

Whether or not records are POD type is computed and recorded in resolution for
all records and can be checked by flags or with a primitive (for use in module
code).

* correctly detect and mark POD types during function resolution
* provide primitive to check if we have a POD type (for bulk transfer opt)
* adjusts bulkCopyRecords to use correct POD detection
* changes uses of FLAG_IGNORE_NOINIT in resolution to use FLAG_NOT_POD because
  overriding the =, init copy, auto copy, or auto destroy generally means that
  you can access the stack trash in a record if no-init is used.
* removed flags FLAG_HAS_USER_ASSIGNMENT, FLAG_HAS_USER_DESTRUCTOR,
  FLAG_HAS_USER_INIT_COPY_FN, FLAG_TRIVIAL_ASSIGNMENT in favor of using the new
  flags FLAG_POD and FLAG_NOT_POD.
* updates pass removeUnnecessaryAutoCopyCalls to use the new mechanism
* adjusts complex2record to mark the new record types with FLAG_POD
* adjusts resolution to mark runtime types as POD or not POD.

Future Work:

* propagateNotPOD normally tries to resolve autoCopy/autoDestroy
  functions, but doing so for runtime types leads to core dumps in the
  current implementation . As a workaround, this patch updates
  propagateNotPOD to avoid trying to resolve autoCopy/autoDestroy for
  runtime types. It is important to understand whether or not runtime
  types should have autoCopy/autoDestroy functions.
* this patch treats dtString and wideStringType specially since the rest
  of the compiler doesn't work correctly if they are POD or not POD.
  Instead, dtString is considered POD during resolution (for the return
  value checking) and then becomes non-POD after that point. Note that
  dtString is a primitive type (not an aggregate type) and wideStringType
  is an aggregate type.  This string workaround can be removed when the
  string type is implemented as a record (and so it was not necessary on
  string-as-rec).

This patch was originally developed on string-as-rec as #2367.  I ported
it over by doing cherry-picks of:

 e0b0859179af0d9e26af644b4196bc1ec69f5a2f
 e56dbc8b3aac40dde3ed285049447f1851a565f8
 4a57298b87339f9a1092e82a3b65b2171d150521
 a097b49b39f99985735bb4ed1fa365af08177b10
 ff16c529f3e2693f268f319e4af240dc167216fb
 fbf4370279e0ff8bf7b609886ed368309f901d74
 4de1de96e7a40081c4d8558e7d9642bdabebc583
 1075ca55621ef8b1a3bd3f3be0f8c9e73215fadc
 199bde94e16cf9cd6cf61b354d8cfa90b89538d8
 659bd0080f9a67584638ccb05462de851b7fb4cd
 909db790f666a7dc951cb955edc0d7007012b891
 95e05bf0e4139795a06003ea6422c21ead31622d
 bbcb76d215d03ae74220bb7e2ae7744009ea33ef
 c0bef3e2282cf4b9ecf6b66af8a09627c7a57c8d
 ff464db3a087a566b59630a1289c24878be34134
 775bb82ea824b3db62168d5a4134b73f237bab76
 733a639092c4cd50ab9bf38724114b864f8f2696
 09debc0f9680fea152b88b7365eb2b3b468c98b9
 9eb245458e57abeb1164e67d7eddffe9af22ba1a
 cf93c3e08291b940802cabd7fd4fe164523704e6

and then updating removeUnnecessaryAutoCopyCalls (since that pass is not
used in string-as-rec), by removing an erroneous change in passByRef, and
adding some fixes for runtime types and strings, which behave differently
on string-as-rec.

Passed full local testing
Passed full GASNet testing
Reviewed by @vasslitvinov.

### CODE REVIEW

**Comment by mppf** [comment]
Local testing underway...

**Comment by mppf** [comment]
Local testing had 5 failures:

[Error matching program output for classes/figueroa/RecordConstructor2]
[Error matching compiler output for compflags/bradc/minimalModules/minModFnRefArg]
[Error matching compiler output for performance/compiler/elliot/passCheck]
[Error matching program output for performance/vectorization/vectorPragmas/forallInStandalone]
[Error matching compiler output for types/type_variables/ferguson/ispod]

**Comment by mppf** [comment]
Running local testing again since the first version ran without running the new code in removeUnnecessaryAutoCopyCalls.

**Comment by mppf** [comment]
Local testing passed.

**Comment by mppf** [comment]
@vasslitvinov, could you give this a final look-over? You already reviewed these changes when they went into string-as-rec, except for the changes to removeUnnecessaryAutoCopyCalls. Nonetheless, I'd appreciate the second set of eyes to make sure I'm only committing what I should be.

File: compiler/optimizations/bulkCopyRecords.cpp [hunk 1]
```diff
@@ -135,11 +44,27 @@ static bool isAssignment(FnSymbol* fn)
...
+    return false;
+
+  // Don't consider assignment for class types to be trivial.
+  // That is because a class type represents a pointer, so
+  // we don't really track whether or not it is POD.
+  // See
+  //   test/classes/hilde/freeAssignment.chpl
+  if (isClass(lhs->type->getValType()))
+    return false;
+
```
Comment (vasslitvinov) [inline compiler/optimizations/bulkCopyRecords.cpp]: IMO this should be either 'return true' or INT_ASSERT(false). Your commit message says that (Chapel) class types ARE pod.

Comment (vasslitvinov) [inline compiler/optimizations/bulkCopyRecords.cpp]: Or do we support user-overloadable assignment etc. on classes, like in freeAssignment? Then it's a different game altogether.

File: compiler/optimizations/bulkCopyRecords.cpp [hunk 2]
```diff
@@ -135,11 +44,27 @@ static bool isAssignment(FnSymbol* fn)
...
+  // Both argument types must be POD types
+  if (isPOD(lhs->type->getValType()) && isPOD(rhs->type->getValType()))
     return true;
```
Comment (vasslitvinov) [inline compiler/optimizations/bulkCopyRecords.cpp]: If the above 'if' is reached, lhs->type->getValType and rhs->type->getValType are equal.
Any reason to test isPOD on both?

File: compiler/include/flags_list.h [hunk 3]
```diff
@@ -105,11 +105,11 @@ symbolFlag( FLAG_GENERIC , npr, "generic" , "generic types, functions and argume
...
-// If any of the following three flags is attached to a class (resp. record or
-// union), then that type is not a Plain-Old Data (POD) type.
-symbolFlag( FLAG_HAS_USER_ASSIGNMENT , npr, "has user assignment" , "applied to classes and records for which assignment is explicitly defined")
-symbolFlag( FLAG_HAS_USER_DESTRUCTOR , npr, "has user destructor" , "applied to classes that define a destructor explicitly" )
-symbolFlag( FLAG_HAS_USER_INIT_COPY_FN , npr, "has user init copy fn" , "applied to classes that define an init copy function explicitly" )
+// Is this type a Plain-Old Data (POD) type - ie no autocopy/destructor/=
+// need ever be called - bit copies will do.
+symbolFlag( FLAG_POD , ypr, "plain old data" , "data can be bit copied")
+symbolFlag( FLAG_NOT_POD , ypr, "not plain old data" , "bit copy overridden")
+
```
Comment (vasslitvinov) [inline compiler/include/flags_list.h]: Please indicate that these two flags can be relied upon only after resolution. (Assuming that's the case.) During resolution, certain function(s) should be called instead. Before resolution, these properties cannot be determined.

File: compiler/passes/flattenFunctions.cpp [hunk 4]
```diff
@@ -93,6 +93,11 @@ passableByVal(Type* type) {
 // Otherwise passing by value is more efficient.
 static bool
 passByRef(Symbol* sym) {
+
+  // Is it a non-POD record type?
+  if( sym->type->symbol->hasFlag(FLAG_NOT_POD) )
+    return true;
+
```
Comment (vasslitvinov) [inline compiler/passes/flattenFunctions.cpp]: I am OK with the above addition going in. However, it seems like it is not the right property to test for.
Please add a comment to the code justifying the relationship of POD-ness to passByRef().

File: compiler/passes/scopeResolve.cpp [hunk 5]
```diff
@@ -1043,6 +1043,12 @@ static void build_constructor(AggregateType* ct) {
       if (method->numFormals() == 2) {
         CallExpr* init = new CallExpr("initialize", gMethodToken, fn->_this);
         fn->insertAtTail(init);
+        // If a type has an initialize method, it's not Plain Old Data.
+        // The only time classes aren't Plain Old Data is when they have the
+        // pragma "no object" associated with them.
+        if (!isClass(ct) || ct->symbol->hasFlag(FLAG_NO_OBJECT)) {
+          ct->symbol->addFlag(FLAG_NOT_POD);
+        }
         break;
```
Comment (vasslitvinov) [inline compiler/passes/scopeResolve.cpp]: Again, I am OK with this going in, however I do not see the relationship of having-object-parent to POD-ness. Please add a comment to the code explaining the relationship.

Furthermore, if a class type can be POD or not, it means (a) we can allow user assignment overloads, and (b) we need to track POD-ness of classes consistently throughout.

Towards that end, I suggest that each Type symbol ends up with exactly one of FLAG_POD and FLAG_NOT_POD coming out of resolution.

File: compiler/optimizations/bulkCopyRecords.cpp (hunk 1 above)
Comment (mppf) [inline compiler/optimizations/bulkCopyRecords.cpp]: I think disabling freeAssignment should be another PR if we do it. I don't think of it as a priority right now, since the optimization in question (bulkCopyRecords) wouldn't do anything useful for class types anyway. I agree that the behavior in that test isn't necessarily what we want to support.

File: compiler/optimizations/bulkCopyRecords.cpp (hunk 2 above)
Comment (mppf) [inline compiler/optimizations/bulkCopyRecords.cpp]: Nope. I don't think it hurts anything either though.

File: compiler/include/flags_list.h (hunk 3 above)
Comment (mppf) [inline compiler/include/flags_list.h]: Sure, I'll add that to the comment.

File: compiler/resolution/functionResolution.cpp [hunk 6]
```diff
@@ -7272,9 +7309,108 @@ static void resolveAutoCopies() {
...
+  AggregateType *at = (AggregateType*) t;
+  bool notPOD = false;
+
+  for_fields(field, at) {
+    Type* ft = field->typeInfo();
+    notPOD |= propagateNotPOD(ft);
+  }
+
+  // make sure we have resolve auto copy/auto destroy.
+  resolveAutoCopy(t);
```
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: above: "resolveD"; capitalize "Make"

File: compiler/resolution/functionResolution.cpp [hunk 7]
```diff
@@ -7272,9 +7309,108 @@ static void resolveAutoCopies() {
...
+  resolveAutoDestroy(t);
+
+  // also check for a non-compiler generated
+  // autocopy/autodestroy.
+  FnSymbol* autoCopyFn = autoCopyMap.get(t);
```
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: Fit the above comment on one line. Capitalize "Also".

File: compiler/resolution/functionResolution.cpp [hunk 8]
```diff
@@ -7272,9 +7309,108 @@ static void resolveAutoCopies() {
...
+
+   After this function is called on a record type, that type and any field
+   record types must have either FLAG_POD or FLAG_NOT_POD set.
+
+   After setting either FLAG_POD or FLAG_NOT_POD if necessary,
+   returns true if FLAG_NOT_POD is set, false otherwise.
+
+   This function should only be called during resolution.
+   Call isPOD (or check FLAG_POD/FLAG_NOT_POD) after resolution.
+ */
```
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: I wonder whether we should stick with the convention of having only // comments.

File: compiler/passes/flattenFunctions.cpp (hunk 4 above)
Comment (mppf) [inline compiler/passes/flattenFunctions.cpp]: I'm adding the comment 
+  // non-POD records are passed by reference here because
+  // other code probably won't copy them correctly if they
+  // are passed by value.

File: compiler/include/flags_list.h (hunk 3 above)
Comment (vasslitvinov) [inline compiler/include/flags_list.h]: Actually, should not rely on these flags at all even post-resolution - should call isPOD(Type*) instead.

File: modules/internal/ChapelBase.chpl [hunk 9]
```diff
@@ -1131,6 +1125,7 @@ module ChapelBase {
     return ir;
   }
   
+  pragma "compiler generated" 
   pragma "donor fn"
   pragma "auto copy fn"
   inline proc chpl__autoCopy(x) return chpl__initCopy(x);
```
Comment (vasslitvinov) [inline modules/internal/ChapelBase.chpl]: Hm, this may break things if chpl__initCopy() is user defined. Right?
Do you know under what conditions this is going to comply with whatever requirements the "compiler generated" pragma promises? Maybe split into two versions, distinguished by where-clause. One version would have that pragma and the other wouldn't?

File: modules/internal/ChapelBase.chpl [hunk 10]
```diff
@@ -1141,7 +1136,11 @@ module ChapelBase {
...
 
   inline proc chpl__autoDestroy(x: object) { }
+
+  pragma "compiler generated" 
   inline proc chpl__autoDestroy(type t)  { }
+
+  pragma "compiler generated" 
   inline proc chpl__autoDestroy(x: ?t) {
     __primitive("call destructor", x);
   }
```
Comment (vasslitvinov) [inline modules/internal/ChapelBase.chpl]: Ditto - this may not necessarily fulfill the requirements of the pragma, depending on what the destructor does?

File: compiler/passes/scopeResolve.cpp (hunk 5 above)
Comment (mppf) [inline compiler/passes/scopeResolve.cpp]: It's actually to handle sync/single variables. I'll update the comment.
I'm not sure if classes can have user assignment overloads. I think that's a design decision we havn't made yet.

I agree about marking each type with FLAG_POD or FLAG_NOT_POD, but I think we have to make some decisions about classes (e.g.) to do that, and would like to put that off for another PR.

I'm adding the comment:

+        // Note: the use of FLAG_NO_OBJECT here is actually intended
+        // to capture sync and single variables. At least it indicates
+        // something is not a normal class instance.

File: compiler/resolution/functionResolution.cpp (hunk 6 above)
Comment (mppf) [inline compiler/resolution/functionResolution.cpp]: OK, done.

File: compiler/resolution/functionResolution.cpp (hunk 7 above)
Comment (mppf) [inline compiler/resolution/functionResolution.cpp]: OK

File: compiler/resolution/functionResolution.cpp (hunk 8 above)
Comment (mppf) [inline compiler/resolution/functionResolution.cpp]: That's not a convention the compiler code is following. e.g. compiler/util/files.cpp. If we decided to (as a team), I'd follow it, but I don't see the point.

File: modules/internal/ChapelBase.chpl (hunk 9 above)
Comment (mppf) [inline modules/internal/ChapelBase.chpl]: What requirements? I thought compiler generated just changed the resolution priority, so that other candidates are preferred to ones marked "compiler generated". If there are other requirements, could you explain?

Comment (vasslitvinov) [inline modules/internal/ChapelBase.chpl]: Ah never mind. There are assumptions about assignment functions when this flag is present. However none of these are assignments.

File: compiler/resolution/functionResolution.cpp (hunk 8 above)
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp]: Sounds good.

File: modules/standard/Types.chpl [hunk 11]
```diff
@@ -130,6 +130,34 @@ The argument must be a type.
...
+  * it has only a compiler generated = routine (when the left hand side
+    and right hand side have the same type. Assignment overloads covering cases
+    where left hand side and right hand side are different types are allowed
+    for POD types)
+  * it contains only POD type fields
+
+Class types in Chapel are always considered POD types (because an instance of
+the class is actually a pointer to the class object, and this pointer is POD).
+
+Primitive numeric Chapel types are POD types as well.
```
Comment (vasslitvinov) [inline modules/standard/Types.chpl]: What about booleans? enums?

Comment (mppf) [inline modules/standard/Types.chpl]: Yes, booleans, enums, c_ptr could be considered POD types. I'd consider booleans and enums "primitive numeric Chapel types". I don't think it really matters though because POD/non-POD is only really interesting for records/tuples. Would you like me to update the comment?

Comment (vasslitvinov) [inline modules/standard/Types.chpl]: Yes, please update the comment.

File: compiler/optimizations/bulkCopyRecords.cpp (hunk 1 above)
Comment (mppf) [inline compiler/optimizations/bulkCopyRecords.cpp]: Vass, I'm removing this check and updating freeAssignment to be a semantic future.

File: compiler/optimizations/bulkCopyRecords.cpp (hunk 2 above)
Comment (mppf) [inline compiler/optimizations/bulkCopyRecords.cpp]: I've updated it to check only one, and include a comment about why that is appropriate.

File: compiler/optimizations/removeUnnecessaryAutoCopyCalls.cpp [hunk 12]
```diff
@@ -371,11 +293,8 @@ isPrimitiveCopy(Vec<Type*>& primitiveCopyTypeSet, Type* type) {
     AggregateType* ct = toAggregateType(type);
     INT_ASSERT(ct);
 
-    // If this class defines an assignment or initCopy or destructor, it is not
-    // POD.
-    if (ct->symbol->hasFlag(FLAG_HAS_USER_DESTRUCTOR) ||
-        ct->symbol->hasFlag(FLAG_HAS_USER_INIT_COPY_FN) ||
-        ct->symbol->hasFlag(FLAG_HAS_USER_ASSIGNMENT))
+    // If we marked this class as not POD, we can't do a primitive copy.
+    if (ct->symbol->hasFlag(FLAG_NOT_POD))
```
Comment (mppf) [inline compiler/optimizations/removeUnnecessaryAutoCopyCalls.cpp]: I've updated this comment to say "marked this type". Note the code refers to everything as "class" because that was the old name in the compiler for AggregateType.

**Comment by mppf** [comment]
Re-running full local testing with the current version.

**Comment by mppf** [comment]
Running testing again...

**Comment by mppf** [comment]
Full local testing passed. I'm going to run a GASNet test too.

**Comment by mppf** [comment]
The current version of this PR has issues multi-locale. 50 failures.
Perhaps we have broken multi-locale strings?

[Error matching program output for classes/hilde/inheritance/fields]
4c4
< Derived foo
---
> D

[Error matching program output for distributions/dm/t2]
[Error matching program output for studies/graph500/kristyn/Graph500_1D_onV/main]
[Error matching program output for studies/graph500/v3/main]
[Error matching program output for studies/hpcc/HPL/vass/bl.md]
[Error matching program output for users/vass/crash3tuple]
(core dump)

[Error matching compiler output for io/ferguson/writef_readf]
[Error matching compiler output for modules/standard/Path/lydia/realpath/fileNotValid]
[Error matching compiler output for modules/standard/Path/lydia/realpath/fileVersion]
[Error matching compiler output for modules/standard/Path/lydia/realpath/onFileInDir]
[Error matching compiler output for modules/standard/Path/lydia/realpath/resultSameAsFilePath]
[Error matching compiler output for multilocale/deitz/needMultiLocales/test_local_string]
[Error matching compiler output for release/examples/benchmarks/shootout/regexdna]
[Error matching compiler output for release/examples/programs/beer]
[Error matching compiler output for studies/590o/blerner/xml-parse]
[Error matching compiler output for studies/beer/beer]
[Error matching compiler output for studies/beer/bradc/beer-promoted-explicit]
[Error matching compiler output for studies/beer/bradc/beer-promoted-init]
[Error matching compiler output for studies/nqueens/nqueens-2-par]
[Error matching compiler output for studies/nqueens/nqueens-2-seq]
[Error matching compiler output for studies/nqueens/nqueens-par]
[Error matching compiler output for studies/nqueens/nqueens-seq]
[Error matching compiler output for studies/shootout/regex-dna/bharshbarg/regexdna]
[Error matching compiler output for trivial/diten/linediff]
> internal error: ALI0069 chpl Version 1.12.0.1d081a8
> Note: This source location is a guess.
>

[Error matching compiler output for studies/amr/advection/amr/AMR_AdvectionCTU_driver]
[Error matching compiler output for studies/amr/advection/grid/Grid_AdvectionCTU_driver]
[Error matching compiler output for studies/amr/advection/level/Level_AdvectionCTU_driver]
[Error matching compiler output for studies/amr/diffusion/grid/Grid_DiffusionBE_driver]
[Error matching compiler output for studies/amr/diffusion/level/Level_DiffusionBE_driver]
?

**Comment by mppf** [comment]
[Error matching program output for distributions/dm/t2]
[Error matching program output for studies/graph500/kristyn/Graph500_1D_onV/main]
still core dumps

**Comment by mppf** [comment]
[Error matching program output for distributions/dm/t2]
[Elapsed time to compile and execute all versions of "studies/graph500/v3/main" - 36.513 seconds]
still core dumps

File: compiler/passes/flattenFunctions.cpp [hunk 13]
```diff
@@ -93,6 +93,14 @@ passableByVal(Type* type) {
...
 static bool
 passByRef(Symbol* sym) {
+
+  // Is it a non-POD record type?
+  // non-POD records/tuples are passed by reference here because
+  // other code probably won't copy them correctly if they
+  // are passed by value.
+  if( isRecord(sym->type) && !isPOD(sym->type) )
+    return true;
+
```
Comment (mppf) [inline compiler/passes/flattenFunctions.cpp]: This seems to be a bad change. It makes test/types/records/ferguson/parallel/coforall.chpl (a string-as-rec test) no longer function, because the iterator yields a reference to r, and r is stored in a heap and freed before the task can launch.

**Comment by mppf** [comment]
Running GASNet testing again...

**Comment by mppf** [comment]
GASNet testing passes!
